File: rate-limiter/rate-limiter.cc
class RateLimiter {
    public:
    /**
     * @param timestamp the current timestamp
     * @param event the string to distinct different event
     * @param rate the format is [integer]/[s/m/h/d]
     * @param increment whether we should increase the counter
     * @return true or false to indicate the event is limited or not
     */
    bool isRatelimited(int timestamp, string& event, string& rate, bool increment) {
        // Write your code here
        stringstream ss;
        ss << rate;
        int limit;
        ss >> limit;
        int base = 1;
        switch(rate[rate.length() - 1]) {
            case 'd': base *= 24;
            case 'h': base *= 60;
            case 'm': base *= 60;
        }
        if (logs.find(event) == logs.end()) {
            logs[event] = vector<int>();
        }
        
        int seg = logs[event].end() - upper_bound(logs[event].begin(), logs[event].end(), timestamp - base);
        if (increment && seg < limit) {
            logs[event].push_back(timestamp);
        }
        return seg >= limit;
    }
    
    private:
    map<string, vector<int> > logs;
};

File: rate-limiter/rate-limiter.py
import bisect

class RateLimiter:
    
    def __init__(self):
        # do some intialize if necessary
        self.logs = dict()
    
    # @param {int} timestamp the current timestamp
    # @param {string} event the string to distinct different event
    # @param {string} rate the format is [integer]/[s/m/h/d]
    # @param {boolean} increment whether we should increase the counter
    # @return {boolean} true or false to indicate the event is limited or not
    def is_ratelimited(self, timestamp, event, rate, increment):
        # Write your code here
        limit, type = rate.split('/')
        limit = int(limit)
        if type == 's': base = 1
        elif type == 'm': base = 60
        elif type == 'h': base = 3600
        elif type == 'd': base = 86400
        
        if event not in self.logs:
            self.logs[event] = []
    
        id = bisect.bisect_right(self.logs[event], timestamp - base)
        
        cnt = len(self.logs[event]) - id
        if increment and cnt < limit:
            self.logs[event].append(timestamp)
return cnt >= limit
